src/resolvers/electrum.rs

```diff
@@ -0,0 +1,161 @@
+use rgbstd::{WitnessAnchor, WitnessId, WitnessOrd, WitnessPos, XAnchor, XPubWitness};
+use std::str::FromStr;
+
+pub struct Resolver {
+    electrum_client: Client,
+    terminal_txes: HashMap<Txid, Tx>,
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Display, Error, From)]
+#[display(doc_comments)]
+pub enum AnchorResolverError {
+    #[from]
+    #[display(inner)]
+    Error(electrum_client::Error),
+
+    /// invalid anchor {0}
+    InvalidAnchor(String),
+}
+
+impl Resolver {
+    #[allow(clippy::result_large_err)]
+    pub fn new(url: &str) -> Result<Self, electrum_client::Error> {
+        let electrum_client = Client::new(url)?;
+        Ok(Self {
+            electrum_client,
+            terminal_txes: none!(),
+        })
+    }
+
+    pub fn add_terminals<const TYPE: bool>(&mut self, consignment: &Consignment<TYPE>) {
+        self.terminal_txes.extend(
+            consignment
+                .terminals
+                .values()
+                .filter_map(|t| t.as_reduced_unsafe().tx.as_ref())
+                .map(|tx| (tx.txid(), tx.clone())),
+        );
+    }
+}
+
+impl ResolveHeight for Resolver {
+    type Error = AnchorResolverError;
+
+    fn resolve_anchor(&mut self, anchor: &XAnchor) -> Result<WitnessAnchor, Self::Error> {
+        let XAnchor::Bitcoin(anchor) = anchor else {
+            panic!("Liquid is not yet supported")
+        };
+        let txid = anchor
+            .txid()
+            .ok_or(AnchorResolverError::InvalidAnchor(format!("{:#?}", anchor)))?;
+
+        if self.terminal_txes.contains_key(&txid) {
+            return Ok(WitnessAnchor {
+                witness_ord: WitnessOrd::OffChain,
+                witness_id: WitnessId::Bitcoin(txid),
+            });
+        }
+
+        fn get_block_height(electrum_client: &Client) -> Result<u64, electrum_client::Error> {
+            Ok(electrum_client.raw_call("blockchain.headers.subscribe", vec![])?["height"]
+                .as_u64()
+                .unwrap())
+        }
+
+        let last_block_height_min = get_block_height(&self.electrum_client)?;
+        let tx_details = self.electrum_client.raw_call(
+            "blockchain.transaction.get",
+            vec![Param::String(txid.to_string()), Param::Bool(true)],
+        )?;
+        let witness_ord = if let Some(confirmations) = tx_details.get("confirmations") {
+            let confirmations = confirmations.as_u64().unwrap();
+            let last_block_height_max = get_block_height(&self.electrum_client)?;
+            let skew = confirmations - 1;
+            let mut tx_height = 0;
+            for height in (last_block_height_min - skew)..=(last_block_height_max - skew) {
+                if let Ok(h) = self.electrum_client.raw_call(
+                    "blockchain.transaction.get_merkle",
+                    vec![Param::String(txid.to_string()), Param::U32(height as u32)],
+                ) {
+                    tx_height = h["block_height"].as_u64().unwrap() as u32;
```

Comment:
Here and everywhere: we must not unwrap, otherwise all wallets can be crashed with a simple change on the electrum server.

Reply:
Correct. Fixed